ui-speedspacechart/src/components/helpers/drawElements/curve.ts

```diff
@@ -1,16 +1,33 @@
 import { clearCanvas, maxPositionValues, speedRangeValues } from '../../utils';
-import type { Store } from '../../../types/chartTypes';
 import { MARGINS } from '../../const';
+import type { DrawFunctionParams, LayerData } from '../../../types/chartTypes';
 
 const { CURVE_MARGIN_TOP, CURVE_MARGIN_SIDES } = MARGINS;
 
-export const drawCurve = (
-  ctx: CanvasRenderingContext2D,
+const drawSpecificCurve = (
+  specificSpeeds: LayerData<number>[],
+  minSpeed: number,
+  speedRange: number,
+  maxPosition: number,
+  ratioX: number,
   width: number,
   height: number,
-  store: Store
+  ctx: CanvasRenderingContext2D
 ) => {
-  const { speed, ratioX, leftOffset } = store;
+  return specificSpeeds.forEach(({ position, value }) => {
+    // normalize speed based on range of values
+    const normalizedSpeed = (value - minSpeed) / speedRange;
+    const x =
+      position.start * ((width - CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
+      CURVE_MARGIN_SIDES / 2;
+    const y = height - normalizedSpeed * (height - CURVE_MARGIN_TOP);
```

Comment:
To improve performance, we can precompute some constants outside the loop to avoid redundant calculations.
```typescript
const adjustedWidth = width - CURVE_MARGIN_SIDES;
const halfCurveMarginSides = CURVE_MARGIN_SIDES / 2;
const adjustedHeight = height - CURVE_MARGIN_TOP;
const xcoef = (adjustedWidth / maxPosition) * ratioX;
```

And then inside the loop, we can reuse it
```typescript
const x = position.start * xcoef + halfCurveMarginSides;
const y = height - normalizedSpeed * adjustedHeight;
```

Reply:
of course yes

Reply:
done